SHA2 extension for RSA signatures

Jakuje opened the pull request.

arisada committed:
bignum: Harmonize ssh_get_random()

Signed-off-by: Aris Adamantiadis <aris@0xbadc0de.be>
Reviewed-by: Andreas Schneider <asn@cryptomilk.org>

arisada committed:
crypto: Split init and finalize functions

Signed-off-by: Aris Adamantiadis <aris@0xbadc0de.be>
Reviewed-by: Andreas Schneider <asn@cryptomilk.org>

arisada committed:
bignum: Make bignum_free safer

Signed-off-by: Aris Adamantiadis <aris@0xbadc0de.be>
Reviewed-by: Andreas Schneider <asn@cryptomilk.org>

Jakuje committed:
client: Properly reset session structures on disconnect

torture_connect_double  test case used to test the connect only up
to key exchange phase, but not after the new keys are stated to be
used for communication. The keys from previous connectoin were not
cleaned up from the previous invocation as well as the seqence
number was not reset and therefore any further packet was failing
with length-check errors or MAC errors.

Signed-off-by: Jakub Jelen <jjelen@redhat.com>
Reviewed-by: Andreas Schneider <asn@cryptomilk.org>

Jakuje committed:
packet: Avoid warnings by using correct template

Signed-off-by: Jakub Jelen <jjelen@redhat.com>
Reviewed-by: Andreas Schneider <asn@cryptomilk.org>

Jakuje committed:
extension negotiation with SHA2 extension for RSA keys

RFC 8308: The extension negotiation in Secure Shell (SSH) Protocol

 * Implementation of client side message requesting the extension
    negotiation and parsing the server answer.

RFC 8332: Use of RSA Keys with SHA-256 and SHA-512
          in the Secure Shell (SSH) Protocol

 * Implementation of client-side parsing of exntesion message used
   to negotiate this extension and allowing the verification of
   host keys signed with this extension.

This intorudces a new key types, that are internally used only for
signature types.

Jakuje committed:
Allow authentication using SHA2 extension to RSA keys

RFC 8332:  Use of RSA Keys with SHA-256 and SHA-512
           in the Secure Shell (SSH) Protocol

 * This change intropduces new API to request signature using a key
   and a different pki mechanism. This is used only for RSA keys,
   that used to have SHA1 hardcoded, but the new algorithsms allow
   to use the SHA2 hashes, if the extension is negotiated.

 * We can not use the ssh_key_alg_to_char() for other keys (ECDSA),
   because they have different  type_c  names while having common
   type id.

Jakuje committed:
SHA2 extension in the ssh-agent interface

The new constants for flags are defined in draft-miller-ssh-agent-02
and active if the SHA2 extension is negotiated with the server.

Jakuje committed:
tests: Verify various host keys can be successfully negotiated and verified

This verifies that all the supported host keys can be used and
verified by the client, including the SHA2 extension in RFC 8332.

Signed-off-by: Jakub Jelen <jjelen@redhat.com>
Reviewed-by: Andreas Schneider <asn@cryptomilk.org>

t8m commented on src/client.c:
```diff
@@ -725,6 +727,10 @@ void ssh_disconnect(ssh_session session) {
+  if (session->next_crypto) {
+    crypto_free(session->next_crypto);
+    session->next_crypto = crypto_new();
+  }
```
You should check NULL return from crypto_new() and call ssh_set_error_oom(session);

Jakuje replied:
Thanks. This should be fixed.

t8m commented on src/kex.c:
```diff
@@ -636,8 +640,19 @@ int ssh_set_client_kex(ssh_session session){
+    snprintf(kex, kex_len, "%s,"KEX_EXTENSION_CLIENT, client->methods[SSH_KEX]);
+    free(client->methods[SSH_KEX]);
+    client->methods[SSH_KEX] = kex;
+
```
Please use realloc instead of malloc+free.

Jakuje replied:
For realloc, I could not use the `snprintf`:
```
/home/jjelen/devel/libssh/src/kex.c: In function ‘ssh_set_client_kex’:
/home/jjelen/devel/libssh/src/kex.c:651:29: warning: passing argument 1 to restrict-qualified parameter aliases with argument 4 [-Wrestrict]
     snprintf(client->methods[SSH_KEX], kex_len, "%s,"KEX_EXTENSION_CLIENT,
              ~~~~~~~~~~~~~~~^~~~~~~~~
              client->methods[SSH_KEX]);
              ~~~~~~~~~~~~~~~~~~~~~~~~
```

t8m replied:
You can just use strcat(client->methods[SSH_KEX], ","KEX_EXTENSION_CLIENT).

Jakuje replied:
True. Thanks for pointer. This works fine.

Jakuje replied:
Fixed now.

t8m commented on src/packet_cb.c:
```diff
@@ -268,3 +268,60 @@ SSH_PACKET_CALLBACK(ssh_packet_service_accept){
+        if (strcmp(name, "server-sig-algs") == 0) {
+            /* TODO check for NULL bytes */
+            SSH_LOG(SSH_LOG_PACKET, "Extension: %s=<%s>",
+                name, ssh_string_get_char(ext_value));
```
You have value already assigned here, why not use it?

t8m commented on src/pki.c:
```diff
@@ -222,6 +224,58 @@ const char *ssh_key_type_to_char(enum ssh_keytypes_e type) {
+  switch (type) {
+    case SSH_KEYTYPE_RSA:
+      if (session->extensions & SSH_EXT_SIG_RSA_SHA512)
+        return SSH_KEYTYPE_RSA_SHA512;
```
Should SHA512 be preferred over SHA256? What does openssh do in this regard?

Jakuje replied:
The openssh stores the indication of the extension in a single variable and if both of them are present, it uses SHA512 (overwriting the value of 256):
https://github.com/openssh/openssh-portable/blob/master/kex.c#L411
To my understanding, RFC does not answer this question so rule of thumb "stronger is better" is used here.

t8m replied:
OK

t8m commented on src/kex.c:
```diff
@@ -636,8 +639,17 @@ int ssh_set_client_kex(ssh_session session){
+    /* Here we append  ext-info-c  to the list of kex algorithsm */
+    kex_len = strlen(client->methods[SSH_KEX]) + strlen(KEX_EXTENSION_CLIENT)
+        + 2; /* comma, trailing zero byte */
+    client->methods[SSH_KEX] = realloc(client->methods[SSH_KEX], kex_len);
```
This is incorrect use of realloc - you need to save the old client->methods[SSH_KEX] value and if the realloc fails, you need to free it, otherwise you leak it. Or you can assign the new value to temporary variable and overwrite the client->methods[SSH_KEX] only if non NULL value is returned.

Jakuje replied:
Thank you. You are right. Temporary variable should do the job here, since on failure, we probably do not want to touch the previously allocated values.

Jakuje committed:
kex: Check allocation result

Signed-off-by: Jakub Jelen <jjelen@redhat.com>
Reviewed-by: Andreas Schneider <asn@cryptomilk.org>

Jakuje committed:
tests: SHA2 extension signatures

This introduces a new test case for RSA unit tests, verifying that
libraries are able to provide and verify the RSA signatures with
SHA2.

Jakuje committed:
kex: The public key algorithms are no longer only host keys only

Signed-off-by: Jakub Jelen <jjelen@redhat.com>

Jakuje committed:
test: Fix test labels (copy & paste error)

Signed-off-by: Jakub Jelen <jjelen@redhat.com>

Jakuje committed:
Configuration PubkeyAcceptedTypes and SSH_OPTIONS_PUBLICKEY_ACCEPTED_TYPES

This adds support for PubkeyAcceptedTypes option in configuration files
and a new option SSH_OPTIONS_PUBLICKEY_ACCEPTED_TYPES having the same
meaning.
This effectively allows to disable using the SHA2 extension, can disable
other old public key mechanisms out of the box (hello DSA) or force
the new SHA2-based key algorithm types if needed.

This exposes the  default_methods  array from  kex.c.

Signed-off-by: Jakub Jelen <jjelen@redhat.com>

Jakuje committed:
session: Do not search for RSA1 keys in ~/.ssh/identity

Signed-off-by: Jakub Jelen <jjelen@redhat.com>

Jakuje committed:
auth: Prevent authentication with non-allowed key algorithms

Signed-off-by: Jakub Jelen <jjelen@redhat.com>

Jakuje committed:
test: Verify the public key algorithms can be limited by configuration option

SSH_OPTIONS_PUBLICKEY_ACCEPTED_TYPES configuration option can limit
what keys can or can not be used for public key authentication.

This is useful for disabling obsolete algorithms while not completely
removing the support for them or allows to configure what public key
algorithms will be used with the SHA2 RSA extension.

Signed-off-by: Jakub Jelen <jjelen@redhat.com>

Jakuje committed:
messages: Generate correct digest for pki signatures

This is no change for old signatures, where the public key algorithm
matches the public key type

This is aproblem when using SHA2 extension for the RSA keys, where
the new signature algorithsm are introduced in addition to the
exitsing ssh-rsa and the signature verification was failing.

Jakuje committed:
pki: Log really the signature algorithm type

Jakuje committed:
server: Support for extension negotiation

This includes intercepting the  ext-info-c  string from
the client kex proposal, configuring the server to allow using
this extension and sending the SSH_MSG_EXT_INFO packet back
to the client after the new keys are in use.

Jakuje committed:
server: We should list SHA2 variants in offered hostkeys

The SHA2 variants should be preferred. Also to buffer needs to be
extended to fit all possible public key algorithms.

cryptomilk committed:
knownhosts: Fix memory leak

CID 1393774

Signed-off-by: Andreas Schneider <asn@cryptomilk.org>

cryptomilk committed:
cmake: Print if server testing is enabled

Signed-off-by: Andreas Schneider <asn@cryptomilk.org>

cryptomilk committed:
libcrypto: Use new RAND_priv_bytes() for strong randomness

Signed-off-by: Andreas Schneider <asn@cryptomilk.org>

cryptomilk committed:
crytpo: Make sure we check return of ssh_get_random() correctly

Signed-off-by: Andreas Schneider <asn@cryptomilk.org>

cryptomilk committed:
torture_pki_dsa: Improve torture_pki_dsa_publickey_base64

Signed-off-by: Andreas Schneider <asn@cryptomilk.org>

cryptomilk committed:
pki: Improve pubkey buffer handling form file

Signed-off-by: Andreas Schneider <asn@cryptomilk.org>

cryptomilk committed:
tests: Fix errno check in pkd

CID 1393877

Signed-off-by: Andreas Schneider <asn@cryptomilk.org>

cryptomilk committed:
tests: Add umask() around mkstemp()

CID 1391450

Signed-off-by: Andreas Schneider <asn@cryptomilk.org>

Jakuje committed:
bind: Complete loading ed25519 in server

Previously, the support was only partial and if the ed25519 key was
the only one, the internal checks were failing the tests.

Jakuje committed:
pkd: Support ed25519 host keys in server

This adds support for the ed25519 keys in the pkd framework and adds
openssh-only tests utilizing these host keys (dropbear does not support
them yet).

Jakuje committed:
pkd: Add missing ECDH mechanisms + whitespace cleanup

Signed-off-by: Jakub Jelen <jjelen@redhat.com>

Jakuje committed:
pki: Support RSA SHA2 extension in server

This involves mostly creation of host keys proofs

TODO: The signing functions could be shared with the client ones
used for public key authentication to avoid code duplication.

Signed-off-by: Jakub Jelen <jjelen@redhat.com>

Jakuje committed:
pkd: Enable the SHA2 extension in server tests

Signed-off-by: Jakub Jelen <jjelen@redhat.com>
